src/rust/lib/graph-editor/src/component/node/port/output.rs

```diff
@@ -0,0 +1,682 @@
+
+        // Case 3: The right circle segment. Always one, if we are in this segment.
+        let case_3_pseudo_bool = in_range_inclusive(position, &middle_segment_end_point, &end);
+        let case_3_value       = case_3_pseudo_bool * 1.0;
+
+        case_2_value + case_3_value
+    }
+
+    /// Compute the crop plane at the location of the given port index. Also takes into account an
+    /// `position_offset` that is given as an offset along the shape boundary.
+    ///
+    /// The crop plane is a HalfPlane that is perpendicular to the border of the shape and can be
+    /// used to crop the shape at the specified port index.
+    fn compute_crop_plane
+    (index:&Var<f32>, port_num: &Var<f32>, width: &Var<f32>, corner_radius:&Var<f32>,
+     position_offset:&Var<f32>) -> AnyShape {
+        let corner_circumference     = corner_radius * 2.0 * PI;
+        let corner_segment_length    = &corner_circumference * 0.25;
+        let center_segment_length    = width - corner_radius * 2.0;
+        let full_shape_border_length = &center_segment_length + &corner_segment_length * 2.0;
+
+        let position_relative = index / port_num;
+        let crop_segment_pos  = &position_relative * &full_shape_border_length + position_offset;
+
+        let crop_plane_pos_relative = calculate_crop_plane_position_relative_to_center_segment(
+            &full_shape_border_length, &corner_segment_length, &crop_segment_pos);
+        let crop_plane_pos          = crop_plane_pos_relative * &center_segment_length
+                                      + corner_radius;
+
+        let plane_rotation_angle = compute_border_perpendicular_angle(&full_shape_border_length,
+                                                                      &corner_segment_length,
+                                                                      &crop_segment_pos);
+        let plane_shape_offset = Var::<Distance<Pixels>>::from(&crop_plane_pos - width * 0.5);
+        let crop_shape         = HalfPlane();
+        let crop_shape         = crop_shape.rotate(plane_rotation_angle);
+        let crop_shape         = crop_shape.translate_x(plane_shape_offset);
+
+        crop_shape.into()
+    }
+
+    ensogl::define_shape_system! {
+        (style:Style, grow:f32, index:f32, port_num:f32, opacity:f32, padding_left:f32,
+        padding_right:f32) {
+            let overall_width  : Var<Distance<Pixels>> = "input_size.x".into();
+            let overall_height : Var<Distance<Pixels>> = "input_size.y".into();
+
+            let base_shape_data = port_base_shape_data(&overall_width, &overall_height, &grow);
+            let BaseShapeData{ port_area,hover_area,radius,width } = base_shape_data;
+
+            let left_shape_crop  = compute_crop_plane(&index,
+                                                      &port_num,
+                                                      &width.clone().into(),
+                                                      &radius.clone().into(),
```

Comment:
can we implement the code without all the explicit `clone().into()` everywhere, please? It would be much nicer to read this way.